File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/SwitchBlocks.java
package com.productions.gizzmoo.pokemonpuzzleleague;

import android.graphics.Point;

import java.io.Serializable;

/**
 * Created by Chrystian on 1/30/2018.
 */

public class SwitchBlocks implements Serializable {

    // 0 based
    private int leftRow;
    private int leftColumn;
    private int rightRow;
    private int rightColumn;

    public boolean switcherIsBeingMoved;

    public SwitchBlocks(int leftC, int leftR, int rightC, int rightR) {
        this.setCoordinates(leftC, leftR, rightC, rightR);
        switcherIsBeingMoved = false;
    }

    public Point getLeftBlock() {
        return new Point(leftColumn, leftRow);
    }

    private boolean isSwitcherStable() {
        return (leftRow == rightRow) && (leftColumn == rightColumn - 1);
    }

    public void setLeftBlock(Point p) {
        if (p.x < 0 || p.x > 4 || p.y < 0 || p.y > 11) {
            throw new RuntimeException("Point is out of bound");
        }

        setCoordinates(p.x, p.y, p.x + 1, p.y);
    }

    public void setRightBlock(Point p) {
        if (p.x < 1 || p.x > 5 || p.y < 0 || p.y > 11) {
            throw new RuntimeException("Point is out of bound");
        }

        setCoordinates(p.x - 1, p.y, p.x, p.y);
    }

    public void setCoordinates(int leftC, int leftR, int rightC, int rightR) {
        leftRow = leftR;
        leftColumn = leftC;
        rightRow = rightR;
        rightColumn = rightC;

        if (!isSwitcherStable()) {
            throw new RuntimeException("Switcher is not stable");
        }
    }

    public boolean areCoordinatesInSwitcher(int x, int y) {
        return areCoordinatesInLeftBlock(x, y) || areCoordinatesInRightBlock(x, y);
    }

    public boolean areCoordinatesInLeftBlock(int x, int y) {
        return leftColumn == x && leftRow == y;
    }

    public boolean areCoordinatesInRightBlock(int x, int y) {
        return rightColumn == x && rightRow == y;
    }

    public void moveUp() {
        if (leftRow > 0 && rightRow > 0) {
            leftRow--;
            rightRow--;
        }
    }

    public void moveDown() {
        if (leftRow < 11 && rightRow < 11) {
            leftRow++;
            rightRow++;
        }
    }

    public boolean isAtTop() {
        return leftRow <= 1;
    }
}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/puzzlegame/SwitchBlocks.kt
package com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame

import android.graphics.Point
import java.io.Serializable

class SwitchBlocks(private var leftColumn: Int, private var leftRow: Int, private var rightColumn: Int, private var rightRow: Int) : Serializable {
    var switcherIsBeingMoved: Boolean = false
    var allowedToBeOnTop: Boolean = true

    init {
        if (!isSwitcherStable) {
            throw RuntimeException("Switcher is not stable")
        }
    }

    val isAtTop: Boolean
        get() = leftRow <= 1

    val leftBlock: Point
        get() = Point(leftColumn, leftRow)

    fun setLeftBlock(p: Point) {
        if (p.x < 0 || p.x > 4 || p.y < 0 || p.y > 11) {
            throw RuntimeException("Point is out of bound")
        }

        setCoordinates(p.x, p.y, p.x + 1, p.y)
    }

    fun setRightBlock(p: Point) {
        if (p.x < 1 || p.x > 5 || p.y < 0 || p.y > 11) {
            throw RuntimeException("Point is out of bound")
        }

        setCoordinates(p.x - 1, p.y, p.x, p.y)
    }

    fun areCoordinatesInSwitcher(x: Int, y: Int): Boolean = areCoordinatesInLeftBlock(x, y) || areCoordinatesInRightBlock(x, y)

    fun areCoordinatesInLeftBlock(x: Int, y: Int): Boolean = leftColumn == x && leftRow == y

    fun areCoordinatesInRightBlock(x: Int, y: Int): Boolean = rightColumn == x && rightRow == y

    fun moveUp() {
        if (leftRow > 0 && rightRow > 0) {
            setCoordinates(leftColumn, leftRow - 1, rightColumn, rightRow - 1)
        }
    }

    fun moveDown() {
        if (leftRow < 11 && rightRow < 11) {
            setCoordinates(leftColumn, leftRow + 1, rightColumn, rightRow + 1)
        }
    }

    private val isSwitcherStable: Boolean
        get() = leftRow == rightRow && leftColumn == rightColumn - 1

    private fun setCoordinates(leftC: Int, leftR: Int, rightC: Int, rightR: Int) {
        leftRow = leftR
        leftColumn = leftC
        rightRow = rightR
        rightColumn = rightC

        if (!isSwitcherStable) {
            throw RuntimeException("Switcher is not stable")
        }
    }
}